enhancements/network/in-cluster-dns-and-loadbalancers-on-more-platforms.md

```diff
@@ -0,0 +1,320 @@
+limitation.
+
+## Proposal
+
+The install-config.yaml platform section for both `external` and `none` will
+include optional fields to deploy and configure coredns and/or the in-cluster
+load balancers. Actual deployment and management will be handled the same way
+it already is on other platforms.
+
+### Workflow Description
+
+A user or automation tool (such as the assisted-installer) that is editing
+install-config.yaml prior to cluster installation will be able to:
+* Enable internal DNS
+* Provide VIPs that implicitly enable in-cluster load balancers
+
+### API Extensions
+
+In the `InstallConfig`, the sections for External and None platforms will have
+new settings that:
+* Enable internal DNS
+* Provide VIPs that implicitly enable internal load balancers (See example under Implementation Details)
+
+The `Infrastructure` API will add fields in the
+[`PlatformSpec`](https://github.com/openshift/api/blob/ef419b6/config/v1/types_infrastructure.go#L272)
+and
+[`PlatformStatus`](https://github.com/openshift/api/blob/ef419b6/config/v1/types_infrastructure.go#L389)
+that mirror the corresponding fields for baremetal, including:
+* `APIServerInternalIPs` in Spec and Status
+* `IngressIPs` in Spec and Status
+* `LoadBalancer` in Status
+
+Those fields will be added to the `External` platform Spec and Status. For the
+`None` platform, a new Spec and Status section will need to be created.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+None
+
+#### Standalone Clusters
+
+The change is only relevant for standalone clusters.
+
+#### Single-node Deployments or MicroShift
+
+Single Node OpenShift benefits from this change as described above. Being a
+single node, it does not need the loadbalancers, but it does require a DNS
+solution.
+
+Assisted-installer already deploys dnsmasq by default as a cluster-internal DNS
+solution for SNO, which has been valuable and successful.
+
+### Implementation Details/Notes/Constraints
+
+The `InstallConfig` will gain new settings for `InClusterLoadBalancer` and
+`InternalDNS`. They are shown below, added to the existing settings for the
+External platform type.
+
+```
+type InClusterLoadBalancer struct {
+    // APIVIPs contains the VIP(s) to use for internal API communication. In
+    // dual stack clusters it contains an IPv4 and IPv6 address, otherwise only
+    // one VIP
+    //
+    // +kubebuilder:validation:MaxItems=2
+    // +kubebuilder:validation:UniqueItems=true
+    // +kubebuilder:validation:Format=ip
+    APIVIPs []string `json:"apiVIPs,omitempty"`
+
+    // IngressVIPs contains the VIP(s) to use for ingress traffic. In dual stack
+    // clusters it contains an IPv4 and IPv6 address, otherwise only one VIP
+    //  
+    // +kubebuilder:validation:MaxItems=2
+    // +kubebuilder:validation:UniqueItems=true
+    // +kubebuilder:validation:Format=ip
+    IngressVIPs []string `json:"ingressVIPs,omitempty"`
+}
```

Comment:
are the [existing API/ingress VIP fields](https://github.com/openshift/installer/blob/cc61cf47b50692bb87cc4d632238d957727aa490/pkg/types/vsphere/platform.go#L76-L106) going to be deprecated/moved to this type?

Reply:
Hopefully that's not necessary. I proposed an API here just to have a standing point for conversation, but I'm happy for the team to weigh in on what would make the most sense. There needs to be some balance of fitting in with the existing APIs, but also making it obvious and simple to enable or disable the feature. And for this proposal, the default should continue to be "disabled" since that matches existing behavior.

Reply:
Following https://github.com/openshift/enhancements/blob/master/enhancements/network/external-lb-vips.md a loadBalancer property was added to the install config [here](https://github.com/openshift/installer/pull/6812/files)
I think this enhancement is the counterpart of https://github.com/openshift/enhancements/blob/master/enhancements/network/external-lb-vips.md and it could probably use the same API